runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/work/processing/failures/WorkFailureProcessor.java

```diff
@@ -0,0 +1,160 @@
+    this.failureReporter = failureReporter;
+    this.heapDumpFetcher = heapDumpFetcher;
+    this.clock = clock;
+    this.retryLocallyDelayMs = retryLocallyDelayMs;
+  }
+
+  /** Returns whether an exception was caused by a {@link OutOfMemoryError}. */
+  private static boolean isOutOfMemoryError(Throwable t) {
+    while (t != null) {
+      if (t instanceof OutOfMemoryError) {
+        return true;
+      }
+      t = t.getCause();
+    }
+    return false;
+  }
+
+  /**
+   * Processes failures caused by thrown exceptions that occur during execution of {@link Work}. May
+   * attempt to retry execution of the {@link Work} or drop it if it is invalid.
+   */
+  public void logAndProcessFailure(
+      String computationId, Work work, Throwable t, Consumer<Work> onInvalidWork) {
+    if (shouldRetryLocally(
+        computationId, work, t instanceof UserCodeException ? t.getCause() : t)) {
+      // Try again after some delay and at the end of the queue to avoid a tight loop.
+      executeWithDelay(retryLocallyDelayMs, work);
+    } else {
+      // Consider the item invalid. It will eventually be retried by Windmill if it still needs to
+      // be processed.
+      onInvalidWork.accept(work);
+    }
+  }
+
+  private String getHeapDumpLog() {
+    return heapDumpFetcher
+        .get()
+        .map(heapDump -> "written to '" + heapDump + "'")
+        .orElseGet(() -> "not written");
+  }
+
+  private void executeWithDelay(long delayMs, Work work) {
+    Uninterruptibles.sleepUninterruptibly(delayMs, TimeUnit.MILLISECONDS);
+    workUnitExecutor.forceExecute(work, work.getWorkItem().getSerializedSize());
+  }
+
+  private boolean shouldRetryLocally(String computationId, Work work, Throwable t) {
+    if (KeyTokenInvalidException.isKeyTokenInvalidException(t)) {
+      LOG.debug(
+          "Execution of work for computation '{}' on key '{}' failed due to token expiration. "
+              + "Work will not be retried locally.",
+          computationId,
+          work.getWorkItem().getKey().toStringUtf8());
+    } else if (WorkItemCancelledException.isWorkItemCancelledException(t)) {
+      LOG.debug(
+          "Execution of work for computation '{}' on key '{}' failed. "
+              + "Work will not be retried locally.",
+          computationId,
+          work.getWorkItem().getShardingKey());
+    } else {
+      LastExceptionDataProvider.reportException(t);
+      LOG.debug("Failed work: {}", work);
+      Duration elapsedTimeSinceStart = new Duration(work.getStartTime(), clock.get());
+      if (!failureReporter.reportFailure(computationId, work.getWorkItem(), t)) {
+        LOG.error(
+            "Execution of work for computation '{}' on key '{}' failed with uncaught exception, "
+                + "and Windmill indicated not to retry locally.",
+            computationId,
+            work.getWorkItem().getKey().toStringUtf8(),
+            t);
+      } else if (isOutOfMemoryError(t)) {
+        LOG.error(
```

Comment:
`getHeapDumpLog` doesn't convey that is is creating the heap dump internally
I think the existing logic is clearer. 

```
          File heapDump = memoryMonitor.tryToDumpHeap();
          LOG.error(
              "Execution of work for computation '{}' for key '{}' failed with out-of-memory. "
                  + "Work will not be retried locally. Heap dump {}.",
              computationId,
              key.toStringUtf8(),
              heapDump == null ? "not written" : ("written to '" + heapDump + "'"),
              t);
```